### ISSUE TITLE
Make `text![["hello", "world"], "bye"]` work?

### ISSUE THREAD

**Issue by kdheepak** [ratatui/ratatui#1831]
When using the macros I'm able to write this:

```rust
        let text = text![line!["hello", "world"], "world"];
```

But I get a compile error when I write this:

```rust
        let text = text![vec!["hello", "world"], "world"];
```

or 

```rust
        let text = text![["hello", "world"], "world"];
```

I thought the individual exprs would be converted to a `Line`, but I get this compile error:

<img width="757" alt="image" src="https://github.com/user-attachments/assets/8b023368-cba3-412e-b2bb-39d0118a453a">

```
error[E0277]: the trait bound `ratatui::prelude::Line<'_>: From<Vec<&str>>` is not satisfied
  --> src/text.rs:48:19
   |
48 |             $line.into(),
   |                   ^^^^ the trait `From<Vec<&str>>` is not implemented for `ratatui::prelude::Line<'_>`, which is required by `Vec<&str>: Into<_>`
...
75 |         let text = text![vec!["hello", "world"], "world"];
   |                    -------------------------------------- in this macro invocation
   |
   = help: the following other types implement trait `From<T>`:
             <ratatui::prelude::Line<'a> as From<&'a str>>
             <ratatui::prelude::Line<'a> as From<String>>
             <ratatui::prelude::Line<'a> as From<Vec<ratatui::prelude::Span<'a>>>>
             <ratatui::prelude::Line<'a> as From<ratatui::prelude::Span<'a>>>
   = note: required for `Vec<&str>` to implement `Into<ratatui::prelude::Line<'_>>`
   = note: this error originates in the macro `text` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `ratatui-macros` (lib test) due to 1 previous error
```

which I was surprised by because I don't get the compile error here:

```rust
        let a = ["hello", "world"];
        let x = Line::from(vec!["hello", "world"]);
        let y = Line::from(a);
        let z: Line = ["hello", "world"].into();
```

Any idea what is going on here? Is there a way to make this work? I tried changing the macro code to  `ratatui::text::Line::from($line),` instead of `$line.into()` and that didn't work either.

**Comment by airblast-dev** [comment]
So I decided to check if this was possible and while possible its pretty ugly.

AFAIK only way to expand the inner array is via TT munching and push down accumulation. I currently have the `text` macro working via such techniques.

If this is something that has interest can someone assign this to me?

**Comment by joshka** [comment]
I don't understand why you wouldn't write:
```rust
text!["hello", "world"]
```
Instead of passing the Vec / array there. Is there a use case where adding the complexity is necessary would be worth considering before implementing that.

**Comment by airblast-dev** [comment]
The idea is we support syntax like this.
```rust
let text = text!["one", ["hello", "bye"], "world"];
assert_eq!(
        text,
        Text::from(vec![
            Line::from("one"),
            Line::from(vec![Span::from("hello"), Span::from("bye")]),
            Line::from("world")
        ])
);
```

I wasn't able to add support for `vec!` as it requires for a hard coded check for the `vec!` macro, but no reason to use it or `line!` with the syntax I mentioned above.

I can create a PR in a bit. If its not wanted we can just close the PR and issue. :)

**Comment by joshka** [comment]
I'm not certain that I like the indirection here enough to make it a thing. Subjectively, the following reads in a way that doesn't really feel all that ergonomic:

```rust
text!["one", ["hello", "bye"], "world"];
```

It's undeniable that this is shorter, but the value you get from a concise expression of syntax should be paired with an obviousness of how that expression should be interpreted. You get that obviousness from drawing similarities from other syntax. Here, this is a nested array with different elements, someone reading the code would generally have to dig into the macro to understand what's happening to the nested elements. It wouldn't be an intuitive reading of the code and raises more questions than it does provide answers.

I feel like the tradeoff here is that we get more concise, but add complexity in a way that burdens the reader rather than the writer of the code. Subjectively, I think this tradeoff falls short of being reasonable. For this reason I'd suggest that this syntax should not be added. I'd listen to other perspectives on this however, particularly if there's more interesting real examples where it makes a lot of sense that this should be supported, but the synthetic ones don't seem compelling to me.

**Comment by airblast-dev** [comment]
That's fair. To be clear I am not advocating for this feature, just wanted to check if its doable as I find it fun to write macros.

> ...someone reading the code would generally have to dig into the macro to understand what's happening to the nested elements
We can provide documentation and examples for it similar to whats done with the example of using `line!` in `text!` but I get your point that just using `line!` is more expressive of whats happening.

**Comment by joshka** [comment]
Let me rephrase that a little as I feel that I misspoke a bit there. Having to look at something other than the code you're reading to discern meaning is the problem, not that you'd have to read the macro's source code. It's more that I enjoy reading code where the intent is clear without having to read documentation.

A perfect simpler example of this problem is that I have a strong bias against accepting boolean parameters in functions. The rationale for this is that without either storing the parameter in a variable, adding a comment, or looking up the docs, someone unfamiliar with the code does not know what it indicates. There are of course exceptions to this rule, but for the most avoiding code where you need to read the docs in order to understand it should be avoided.

Here that same ideas would happen when reading code that uses this macro with the extra syntax. Seeing a nested array and intuitively understanding and knowing that the macro is picking up the array syntax and treating that specially rather than the rust language treating it as just an array is where the problematic part of this is. I don't think adding `line!` in front of that part of some array seems particularly bothersome to the point where I'd want to provide that ambiguity.
